packages/wonder-blocks-core/components/with-ssr-placeholder.js

```diff
@@ -0,0 +1,158 @@
+ * the component tree. This will also be the case across portal
+ * boundaries.
+ *
+ * Example:
+ *
+ * ```js
+ * <WithSSRPlaceholder placeholder={() => <div>Renders on the server!</div>}>
+ *   {() => <div>Only renders on the client (after rehydration).</div>}
+ * </WithSSRPlaceholder>
+ * ```
+ */
+export default class WithSSRPlaceholder extends React.Component<Props, State> {
+    state = {
+        mounted: false,
+    };
+
+    componentDidMount() {
+        if (this._isTheRootComponent) {
+            // We only want to force a new render if we were responsible for
+            // the first render, so we guard that state change here.
+
+            // eslint-disable-next-line react/no-did-mount-set-state
+            this.setState({
+                mounted: true,
+            });
+        }
+    }
+
+    _isTheRootComponent = false;
+
+    _renderAsRootComponent() {
+        const {mounted} = this.state;
+        const {children, placeholder} = this.props;
+
+        // We are the first component in the tree.
+        // We are in control of instigating a second render for our
+        // component tree.
+        this._isTheRootComponent = true;
+
+        if (mounted) {
+            // This is our second non-SSR render, so let's tell everyone to
+            // do their thing.
+            return (
+                <RenderStateContext.Provider value="standard">
+                    {children()}
+                </RenderStateContext.Provider>
+            );
+        }
+
+        // OK, this is the very first render.
+        // If we have a placeholder, we render it, and ensure that any
+        // nested SSR components know we're still on that first render
+        // but they're not in charge of instigating the second render.
+        if (placeholder) {
+            return (
+                <RenderStateContext.Provider value="initial">
+                    {placeholder()}
+                </RenderStateContext.Provider>
+            );
+        }
+
+        // Otherwise, we return nothing.
+        return null;
+    }
+
+    _maybeRender(renderState: RenderState) {
+        const {children, placeholder} = this.props;
+
+        switch (renderState) {
+            case "root":
+                return this._renderAsRootComponent();
+
+            case "initial":
+                // We're not the root component, so we just have to either
+                // render our placeholder or nothing.
+                // The second render is going to be triggered for us.
+                if (placeholder) {
+                    return placeholder();
+                }
+                // Otherwise, we render nothing.
+                return null;
+
+            case "standard":
+                // We have covered the SSR render, we're now rendering normally.
+                return children();
+
+            default:
+                throw new Error("We got a render state we don't understand");
```

Comment:
Needs a test case

Reply:
I hadn't considered excluding exceptions from code coverage.  This is one of those things that should in theory never happen, right?

Reply:
Yes, it's there for us when we're maintaining code to help avoid hidden issues.